java/ql/src/semmle/code/java/frameworks/ratpack/Ratpack.qll

```diff
@@ -0,0 +1,101 @@
+/**
+ * Provides classes and predicates related to `ratpack.*`.
+ */
+
+import java
+private import semmle.code.java.dataflow.DataFlow
+private import semmle.code.java.dataflow.FlowSteps
+
+/**
+ * Ratpack methods that access user-supplied request data.
+ */
+abstract class RatpackGetRequestDataMethod extends Method { }
+
+/**
+ * The interface `ratpack.http.Request`.
+ * https://ratpack.io/manual/current/api/ratpack/http/Request.html
+ */
+class RatpackRequest extends RefType {
+  RatpackRequest() {
+    hasQualifiedName("ratpack.http", "Request") or
+    hasQualifiedName("ratpack.core.http", "Request")
+  }
+}
+
+/**
+ * Methods on `ratpack.http.Request` that return user tainted data.
+ */
+class RatpackHttpRequestGetMethod extends RatpackGetRequestDataMethod {
+  RatpackHttpRequestGetMethod() {
+    getDeclaringType() instanceof RatpackRequest and
+    hasName([
+        "getContentLength", "getCookies", "oneCookie", "getHeaders", "getPath", "getQuery",
+        "getQueryParams", "getRawUri", "getUri"
+      ])
+  }
+}
+
+/**
+ * The interface `ratpack.http.TypedData`.
+ * https://ratpack.io/manual/current/api/ratpack/http/TypedData.html
+ */
+class RatpackTypedData extends RefType {
+  RatpackTypedData() {
+    hasQualifiedName("ratpack.http", "TypedData") or
+    hasQualifiedName("ratpack.core.http", "TypedData")
+  }
+}
+
+/**
+ * Methods on `ratpack.http.TypedData` that return user tainted data.
```

Comment:
According to @ldaley, one of the Ratpack authors, almost always.

The actual source of the data would be `Request::getBody`, where `TypedData` is where this input gets pulled from the body. However, there are cases in user libraries where `UploadedFile` (a supertype of `TypedData`) is instantiated inside a Jackson deserialized object, which isn't otherwise easily traceable.

Now that you're asking, it would probably be better to use `TypedData` as a taint passthrough and make sure that `getBody` is captured as a source instead. 🤔